java/ql/src/Telemetry/AutomodelFrameworkModeCharacteristics.qll

```diff
@@ -14,23 +14,10 @@ private import semmle.code.java.Expr as Expr
 private import semmle.code.java.security.QueryInjection
 private import semmle.code.java.security.RequestForgery
 private import semmle.code.java.dataflow.internal.ModelExclusions as ModelExclusions
+private import AutomodelSharedUtil as AutomodelSharedUtil
 import AutomodelSharedCharacteristics as SharedCharacteristics
 import AutomodelEndpointTypes as AutomodelEndpointTypes
 
-/**
- * A meta data extractor. Any Java extraction mode needs to implement exactly
- * one instance of this class.
- */
-abstract class MetadataExtractor extends string {
-  bindingset[this]
-  MetadataExtractor() { any() }
-
-  abstract predicate hasMetadata(
-    DataFlow::ParameterNode e, string package, string type, boolean subtypes, string name,
-    string signature, int input, string parameterName
-  );
-}
-
```

Comment:
We've discussed this a bit during development. Having an interface that both metadata extractors could use is easy:

```
abstract predicate hasMetadata(Endpoint e, string name, string value)
```

That was what the first version of the class looked like in the framework mode PR. There, the feedback was to not use that, but pick an interface where the type check catches typos.

What we could do here would be to have an abstract superclass that has

```
abstract predicate hasMetadata(
    DataFlow::ParameterNode e, string package, string type, boolean subtypes, string name,
    string signature, int input, /* skip that one string parameterName */
  );
```

but we'd still need to have the parameter name in an extra tacked on method. This would be quite awkward.

Perhaps we could do also some advanced construction that's leveraging the parameterized modules language feature, but that would be cracking nuts with a sledgehammer and be more difficult to read IMHO..

LMK what you think.

Reply:
Any thoughts, @adityasharad?

Reply:
I don't fully understand, which means I need to look at the two implementations more closely and maybe chat with you.

Totally agree the `(Endpoint e, string name, string value)` form would be difficult to type check, let's avoid that.
I was hoping for something like `abstract predicate hasMetadata(Endpoint e, /* all other parameters here */);

Is the challenge that you have a slightly different set of predicate parameters in the two modes, with `parameterName` only present in one mode?

Reply:
> Is the challenge that you have a slightly different set of predicate parameters in the two modes, with parameterName only present in one mode?

Yes, indeed. Metadata is in general mode-specific, but there's a shared set of metadata for (package, type, name, ...)

Reply:
I wonder if this is best left for a follow-up PR? This PR is getting to be a bit big, and I think we've addressed all the other outstanding issues. 🙂

Reply:
Yeah let's design separately and create the common interface in a follow-up PR.